[#69095] Primerize Admin > System Settings forms

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/69095

# What are you trying to accomplish?

This PR converts multiple admin system settings views from legacy forms to Primer-based forms, specifically targeting repositories, languages, general, and experimental settings pages. The changes modernize the UI to use GitHub's Primer Design System via ViewComponent while maintaining existing functionality.

- Migrates settings forms from `styled_form_tag` to `settings_primer_form_with` with corresponding Primer form components
- Extracts form logic into dedicated form classes (e.g., `Admin::Settings::RepositoriesSettingsForm`, `GeneralSettingsForm`)
- Adds new input methods to Settings::InputMethods including `text_area`, `rich_text_area`, `select_list`, and `check_box_group`.

## Screenshots

### General

<img width="990" height="773" alt="Screenshot 2025-11-25 at 19 33 22" src="https://github.com/user-attachments/assets/8c6ec77a-d9be-4d73-8e54-2fc825c1a8c8" />

<img width="1032" height="769" alt="Screenshot 2025-11-25 at 19 33 36" src="https://github.com/user-attachments/assets/4723a549-34d5-41b3-abf1-37fa50509d84" />

### Languages

<img width="988" height="772" alt="Screenshot 2025-11-25 at 19 33 46" src="https://github.com/user-attachments/assets/38c23f33-41ea-4489-b208-70552d4ed505" />

### Repositories

<img width="981" height="776" alt="Screenshot 2025-11-25 at 19 33 56" src="https://github.com/user-attachments/assets/c35af51f-0eea-44d2-a89b-75c1130fd0c8" />

<img width="1001" height="771" alt="Screenshot 2025-11-25 at 19 34 06" src="https://github.com/user-attachments/assets/3eddff13-9833-4bcb-9e24-bbf4f87d4d4d" />

### Feature flags

<img width="993" height="769" alt="Screenshot 2025-11-25 at 19 34 27" src="https://github.com/user-attachments/assets/744efd33-eb5a-400e-8c6e-9e868ff8cadc" />

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
Fix calling instance methods from settings_form

`settings_form` was previously executing the block while still in the
class context. `instance_exec` ensures the block runs with `self` set to
the form instance. As such `settings_form` now behaves like its standard
`form` counterpart, i.e. the following works as expected:

    class MySettingsForm < ApplicationForm
      settings_form do |f|
        f.text_field(name: :my_setting, caption: text_field_caption)  # now resolves correctly
      end

      private

      def text_field_caption
        "A long caption"
      end
    end

Also removes `render_inline_form` in favour of the namesake method
provided by upstream Primer.

myabc commented:
@HDinger this is ready for some initial feedback, particularly on the visuals. I still have a it of tidying up and possibly some test fixing to do..

HDinger reviewed (commented):
I did only review visually until now and did not check the code yet but I have to say, it looks really great! Well done 👏 🌟 
I only found one little detail which I stumbled over: In the repository settings, the API key field is somehow indented to the right. It also feels as if the spacing is a bit too tight, but I might be wrong about that.

<img width="521" height="273" alt="Image" src="https://github.com/user-attachments/assets/f1c61d28-aa43-4116-86d6-26603cab53c8" />

myabc commented:
>  In the repository settings, the API key field is somehow indented to the right. It also feels as if the spacing is a bit too tight, but I might be wrong about that.

@HDinger the indent is something that Primer does by default when using nested forms and checkboxes. This came up when I worked on the _Login and registration page_ (PR #20846) and in that case, I added some additional styling to override and remove the indent. However I'd prefer to discuss before overriding here.

There's an Open Point to discuss here: https://community.openproject.org/wp/68696

I agree about the top margin of API key field.

HDinger commented on app/forms/admin/settings/general_settings_form/allowed_link_protocols_caption.html.erb:
```diff
@@ -0,0 +1,12 @@
+      mailto_code: content_tag(:code, "mailto")
+    ).html_safe
+  %>
+</div>
```
I admit, that I don't fully understand how this works, but it feels super weird. This is a completely new pattern and I don't see the reason for it. There is also already a `setting_caption` helper. Why don't you use that one?

myabc replied:
I guess these are exceptions to the rule - this is the same I18n string that the previous implementation rendered.

The fact that this translation uses interpolation makes it harder to use the automatic `setting_caption` support.

HDinger commented on app/forms/admin/settings/repositories_settings/checkout_form.rb:
```diff
@@ -0,0 +1,91 @@
+            }
+          ) do |setting_repository_checkout_check_box|
+            setting_repository_checkout_check_box.nested_form(
+              classes: ["mt-2", { "d-none" => Setting.repository_checkout_data[vendor.to_s]["enabled"].to_i.zero? }],
```
I am not a fan of using these utility classed directly. Primer does sometimes change them.. Why don't you pass that as normal system arguments? As far as I can tell `nested_form` can deal with normal attributes.

myabc replied:
`nested_form` doesn't handle system arguments unfortunately 😒

HDinger replied:
Too bad :(

HDinger commented on app/forms/admin/settings/repositories_settings/commit_messages_form.rb:
```diff
@@ -0,0 +1,105 @@
+            }
+          ) do |commit_logtime_check_box|
+            commit_logtime_check_box.nested_form(
+              classes: ["mt-2", { "d-none" => !Setting.commit_logtime_enabled? }],
```
Same as  before

myabc replied:
We can switch to using `hidden` - since it's a standardized attribute, it is less likely to break.

HDinger commented on app/forms/settings/input_methods.rb:
```diff
@@ -154,10 +220,97 @@ def multi_language_text_select(name:, current_language: I18n.locale.to_s)
+
+        args[:label] ||= setting_label(name, args[:value])
+        args[:caption] ||= setting_caption(name, args[:value])
+        args[:checked] = setting_value(name).include?(args[:value]) unless args.key?(:checked)
```
Isn't that the excat same thing as for the radio buttons? If yes, maybe extracting it, could increase readability

myabc replied:
@HDinger yep, I'm not very happy with this code - we're basically reimplementing `[options_for_select](https://api.rubyonrails.org/classes/ActionView/Helpers/FormOptionsHelper.html#method-i-options_for_select)`, but with sane defaults for settings.

myabc replied:
@HDinger ok, I've tried to tidy up the value processing. It's still ugly, but at least it's tested.

HDinger reviewed (changes requested):
Code-wise, I have the biggest question marks with the `..caption.html.erb` files. This is a new pattern, and at least for me, a hard to understand pattern. It also invites people to add more HTML in the caption while we usually go for one translated (and maybe formatted) string. Maybe I am missing some key aspect of that implementation but I am really not a fan of that.

HDinger commented:
Ahh.. now I understand, you are using the [caption template pattern](https://primer.style/view-components/lookbook/pages/forms/caption_templates/) from primer.. Well, that makes things clearer.. It also 
explains why I could not find any reference in code 😅 Maybe it is then worth adding some inline comments somewhere? 🤔

myabc commented:
@copilot please try to try up the values handling in input_methoda.rb

myabc committed:
Introduce new settings input methods

Introduces `text_area`, `check_box_group` and `select_list` inputs.
Makes various refactorings to ensure that handling of explicit `valus`
params is as robust as possible.

myabc committed:
[#69095] Primerize Admin > System Settings forms

myabc commented on app/forms/admin/settings/repositories_settings/commit_messages_form.rb:
```diff
@@ -0,0 +1,105 @@
+              name: :commit_fix_status_id,
+              values: available_statuses,
+              label: I18n.t(%i[setting_commit_fix_keywords label_applied_status]).join(": "),
+              prompt: "--- #{I18n.t(:actionview_instancetag_blank_option)} ---"
```
I find this old-fashioned prompt style ugly FWIW.

myabc commented on app/forms/admin/settings/repositories_settings/commit_messages_form.rb:
```diff
@@ -0,0 +1,105 @@
+          sf.group(layout: :horizontal) do |group|
+            group.text_field(
+              name: :commit_fix_keywords,
+              label: I18n.t(%i[setting_commit_fix_keywords label_keyword_plural]).join(": "),
```
I find it annoying that there isn't a way to add a group label (or fieldset with legend) in Primer Forms. Concatenating the labels is my work-around.

myabc commented on app/forms/admin/settings/experimental_settings_form.rb:
```diff
@@ -0,0 +1,62 @@
+      private
+
+      def available_feature_flags
+        OpenProject::FeatureDecisions
+          .all
+          .map { [it.to_s.humanize, "feature_#{it}_active"] }
```
Our Feature Flags implementation is pretty bare-boned. Common functionality like getting a humanised name or looking up the Setting name should be moved there IMO.

myabc committed:
Use aria-describedby to associate field units

myabc commented on spec/support/forms/rendered_inline_settings_form.rb:
```diff
@@ -0,0 +1,42 @@
+        render_inline_settings_form(f, &)
+      end
+    end
+  end
```
It's in a different scope, so it works... I can't think of a better name at the moment!

HDinger commented on app/components/op_primer/form_helpers.rb:
```diff
@@ -30,51 +30,23 @@
-    # @param blk [Proc] A block that defines the form structure.
-    def render_inline_form(form_builder, &blk)
-      form_class = Class.new(ApplicationForm) do
-        # This is a workaround to make the form class aware of the template path
-        # of the block that is passed to it.
-        #
-        # This avoids an annoying warning "Could not identify the template" from
-        # original `base_template_path` method.
-        define_singleton_method(:base_template_path) do
-          blk.source_location.first
-        end
-
-        form(&blk)
```
I was not aware that Primer has a `render_inline_form` itself, TIL. Looking at it, the implementation is however different, and I am wondering wether this part here might still be needed. Wdyt?

myabc replied:
I'll check the CI output for warnings.

myabc replied:
will also ping @cbliard, who may know more.

myabc replied:
I haven't seen any warnings in the logs.

HDinger commented on spec/support/forms/rendered_inline_form.rb:
```diff
@@ -0,0 +1,42 @@
+
+  let(:form_arguments) { { url: "/foo", model: } }
+
+  def render_inline_form(&)
```
As far as I can see, this example is not used at all. I admit that it is rather annpying, that the IDE identifies this method as the definition of `render_inline_form` (and not the Primer one) and thus always shows an error. So maybe it is better to remove this file.

<img width="555" height="183" alt="Image" src="https://github.com/user-attachments/assets/f21b6fe0-9322-4155-84d1-0da4f2f72888" />

myabc replied:
I'll fix this.

myabc replied:
> As far as I can see, this example is not used at all.

This shared context isn't used yet. It was added for symmetry with the other shared contexts.

HDinger commented on app/forms/settings/input_methods.rb:
This file is still a monster and hard to understand if you were not involved. However, I have no ideas for improvement, and it was already very complex before. 🤷

myabc replied:
I would actually like to refactor this further, but it will have to come in a later PR. I think that decorating existing inputs (particularly `select_list`, `radio_button_group` and `check_box_group`) adds some additional complexity so we might want to go with a different approach.

myabc committed:
Fix naming collision in with *form shared contexts

Prefixes methods with `vc_` - similar to convention used in
`ViewComponent::TestHelpers`.

See https://github.com/ViewComponent/view_component/blob/main/lib/view_component/test_helpers.rb#L245
